README.md
# _WeNet Telegram Chatbots_

:warning: **Not the 100% of the application is open source. In fact, some of the required libraries are private of U-Hopper. These chatbot applications should be intended only for example on how to build an application using the WeNet platform.**

## Introduction

These chatbots are created in the context of `WeNet: the Internet of us`.

These Telegram chatbots are the first WeNet applications, allowing users to communicate with each other to organize shared meals with potential unknown people, or to ask questions and generic help, in a strong privacy-oriented manner.

The _eat-together_ bot allows users to create a new _task_, representing a shared meal. The WeNet platform will select some users potentially interested in the newly created task, and the bot will ask them to candidate to participate to it. Subsequently, the task creator will be notified of a candidature, being asked whether to accept or reject it. Candidates will acknowledge the owner's decision. Eventually, the task is closed by the owner using the chatbot, providing also an outcome, either successfully or failed.

The _ask-for-help_ bot allows to create _tasks_ that are questions, to which other WeNet users will answer.

The bots expose an HTTP endpoint to receive messages and notifications from the WeNet platform.

## Documentation

### How the bots work? High-level overview

The bots are _deterministic finite automatas_ (DFA), that are graphs where each node is a possible _state_ in which the conversation can be, and the edges are the _transactions_ between two states. For example:
1. The bot is in its initial state `S0`, which means that the user has not used it yet.
2. The user uses the `/start` command: this is a transaction, going from the initial state `S0` to the next state `S1`.
3. State `S1` proposes a message with 3 buttons: `B1`, `B2` and `B3`; each button makes the state change to a different state.
4. So we can have 3 transactions exiting from `S1`: the one triggered by `B1`, the one triggered by `B2` and the one triggered by `B3`.
5. Each transaction makes the user land into a state, either `S2`, `S3` or `S4`.

So the next state `S_(i + 1)` depends from the current state `S_(i)` and from the input received by the bot.

#### How the DFA is implemented?

The `WenetEventHandler` is an abstract class containing the common things used by the two _real_ handlers of the two bots, including the management of the DFA. 

The current state is saved in the user context, using as key the constant `self.CONTEXT_CURRENT_STATE`, while the value represents the state itself. Of course the value must be unique for each possible state of the bot, so it can be useful to define each possible state as constant in the handler class.

Every exchange of message between the user and the bot implies a change of state, so every time the context must be updated. 
Any possible additional piece of information needed to continue the conversation flow must be saved in the context. 
For example, to create a new taks the user must specify several information (a date, a title, a description, etc): 
all these data points are collected in several steps, each one represented by a state, and saving every time the user's inputs in the context. 
Only in the final state, the task is created and saved.

Assuming that we are in a state where we are expecting that the user types the name of the task, we know that the next input coming from the user will be the name of the task, and so we can use the intent manager to route the flow of the bot.
For example, the following piece of code calls the function `self.organize_q2` passing to it the intent `self.ORGANIZE_Q2`, and calling it every time an input is received and the current state is equal to `self.ORGANIZE_Q1`.
```python
self.intent_manager.with_fulfiller(
    IntentFulfillerV3(self.ORGANIZE_Q2, self.organize_q2).with_rule(
        static_context=(self.CONTEXT_CURRENT_STATE, self.ORGANIZE_Q1))
)
```

### Messages from Wenet

The bots can receive messages from the Wenet platform (notifications, textual messages, etc). Some methods are already available in the wenet handler to manage these situations:
- `handle_wenet_textual_message()` is triggered every time a `TextualMessage` is sent to the bot;
- `handle_wenet_authentication_result()` is triggered every time an user performs the authentication in the Wenet Hub;
- `handle_wenet_message()` is triggered in all the remaining cases, so when the message from Wenet is neither a `TextualMessage` nor a `WeNetAuthenticationEvent`; each application implements its own custom messages, and this is the place to handle them.

In these cases the messages are sent directly to the bot, without passing through the chatbot interface. So the user context must be explicitly fetched, together with the user information:
1. Get the user account related with the receiver id of the message:
```python
user_accounts = self.get_user_accounts(message.receiver_id)
if len(user_accounts) != 1:
    raise Exception(f"No context associated with Wenet user {message.receiver_id}")

user_account = user_accounts[0]
```
2. Get the context (it comes for free):
```python
context = user_account.context
```
3. Instanciate the service API:
```python
service_api = self._get_service_api_interface_connector_from_context(context)
```
The service API handler manages the OAuth tokens, so every time it is invoked is safe to call the following instruction to save the updated token:
```python
context = self._save_updated_token(context, service_api.client)
```
4. At the end, a `NotificationEvent` must be returned:
```python
return NotificationEvent(user_account.social_details, response_list, context)
```
where `response_list` is a list of `ResponseMessage` objects.

## Setup and configuration

### Installation

The chatbot required Python version 3.7 or higher.

All required Python packages can be installed using the command:

```bash
pip install -r requirements.txt
```

### Docker support

A dedicated Docker image for this component can be build by taking advantage of the repository Docker support.
The command:
```bash
./build_docker_image.sh
```
will:

* run the tests on the checked-out version of the service APIs;
* build the docker image for the chatbot (the naming is the following `registry.u-hopper.com/`)

## Usage

In order to run the _eat-together_ chatbot, run the following command:
```bash
python -m eat_together_bot.main
```

In order to run the _ask-for-help_ chatbot, run the following command:
```bash
python -m ask_for_help_bot.main
```

In order to run the endpoint, run the following command:
```bash
python -m messages.main
```

### Chatbot env variables

Both the chatbots use the following environment variables:
* `TELEGRAM_KEY`: secret key provided by Telegram to use the bot API
* `MQTT_HOST`: MQTT host
* `MQTT_SUBSCRIBER_ID`: MQTT subscriber id for the client
* `MQTT_USER`: MQTT user
* `MQTT_PASSWORD`: MQTT password
* `MQTT_TOPIC`: MQTT topic to listen on
* `INTERFACE_APIKEY`: bot interface api key
* `INSTANCE_NAMESPACE`: bot instance namespace
* `WENET_INSTANCE_URL`: the url of the WeNet instance
* `WENET_APP_ID`: WeNet App ID used by the bot
* `WENET_HUB_URL`: url of the WeNet hub
* `TASK_TYPE_ID`: the type ID of the tasks
* `COMMUNITY_ID`: the community ID
* `CLIENT_SECRET`: the secret key of the WeNet application
* `WENET_AUTHENTICATION_URL`: the URL that manages authentication in WeNet
* `WENET_AUTHENTICATION_MANAGEMENT_URL`: the URL that manages OAuth in WeNet
* `REDIRECT_URL`: the redirection URL associated with the WeNet application
* `PROJECT_NAME` (optional): a string that will be used as name of the log file (with the format `<PROJECT_NAME>.log`). The default value is `wenet-ask-for-help-chatbot`
* `LOCALE_TTL` (optional): the time to live of the Redis key in which the user locale is saved, in seconds. By default, it is 86400 (24h).
* `SENTRY_DSN`: (Optional) The data source name for sentry, if not set the project will not create any event
* `SENTRY_RELEASE`: (Optional) If set, sentry will associate the events to the given release
* `SENTRY_ENVIRONMENT`: (Optional) If set, sentry will associate the events to the given environment (ex. `production`, `staging`)

The ask4help bot also needs the following environment variables:
* `MAX_USERS`: the maximum number of users that should receive the question in the ask4help bot (default is 5)
* `MAX_ANSWERS`: the maximum number of answers that should receive a question in the ask4help bot before expiring (default is 15)
* `EXPIRATION_DURATION`: the maximum amount of time before a question in the ask4help bot expires (default is 86400)
* `NEARBY_EXPIRATION_DURATION`: the maximum amount of time before a nearby question in the ask4help bot expires (default is 7200)
* `SURVEY_URL`: the url of the survey
* `PILOT_HELPER_URL`: (Optional) the url of the helper page specific for a pilot
* `CHANNEL_ID`: (Optional) the id of the channel where to publish questions and best answers
* `PUBLICATION_LANGUAGE`: (Optional) the language in which to publish messages on the channel. The default is `en`

For the translations of the badges messages the following environment variables are needed:
* `FIRST_QUESTION_BADGE_ID`: the id of the first question badge
* `CURIOUS_LEVEL_1_BADGE_ID`: the id of the curious level 1 badge
* `CURIOUS_LEVEL_2_BADGE_ID`: the id of the curious level 2 badge
* `FIRST_ANSWER_BADGE_ID`: the id of the first answer badge
* `HELPER_LEVEL_1_BADGE_ID`: the id of the helper level 1 badge
* `HELPER_LEVEL_2_BADGE_ID`: the id of the helper level 2 badge
* `FIRST_GOOD_ANSWER_BADGE_ID`: the id of the first good answer badge
* `GOOD_ANSWERS_LEVEL_1_BADGE_ID`: the id of the good answerer level 1 badge
* `GOOD_ANSWERS_LEVEL_2_BADGE_ID`: the id of the good answerer level 2 badge
* `FIRST_LONG_ANSWER_BADGE_ID`: the id of the first long answer badge
* `EXPLAINER_LEVEL_1_BADGE_ID`: the id of the explainer level 1 badge
* `EXPLAINER_LEVEL_2_BADGE_ID`: the id of the explainer level 2 badge

Optional variables (to setup Redis):
- `REDIS_HOST` (default is `localhost`)
- `REDIS_PORT` (default is 6379)
- `REDIS_DB` (default is 0)

The _ask for help_ bot has the following optional environment variable:
* `TRANSLATION_FOLDER_PATH`, indicating the path of the folder in which translations are stored (default is `../../translations`).

### Endpoint env variables

The endpoint uses the following environment variables:
* `MESSAGES_HOST`: host running the APIs (default to `0.0.0.0`)
* `MESSAGES_PORT`: port of the host (default to `12345`)
* `MQTT_HOST`: MQTT host
* `MQTT_PUBLISHER_ID`: MQTT publisher id for the client
* `MQTT_USER`: MQTT user
* `MQTT_PASSWORD`: MQTT password
* `MQTT_TOPIC`: MQTT topic to write on
* `INSTANCE_NAMESPACE`: bot instance namespace
* `WENET_APP_ID`: WeNet App ID used by the bot
* `WENET_HUB_URL`: url of the WeNet hub
* `BOT_ID`: the bot ID associated with the EventHandler used by the bot itself.
* `PROJECT_NAME` (optional): a string that will be used as name of the log file (with the format `<PROJECT_NAME>-messages.log`). The default value is `wenet-ask-for-help-chatbot`.

## Contributing

Contributions to this project are more than welcome.
Contributions regarding the supported languages is particularly appreciated.
Details about the contribution guidelines can be found in [CONTRIBUTING.md](CONTRIBUTING.md).

## Instances

The production instance of the ask4help chatbot is available here [https://t.me/wenet_ask_for_help_bot](https://t.me/wenet_ask_for_help_bot).


CHANGELOG.md
# Changelog

## Version 3.*

### 3.3.3

:nail_care: Polish
- Updated mongolian translation for social closeness question.

### 3.3.2

:nail_care: Polish
- Updated english translation for social closeness question.
- Refactor of the button for social closeness question. 

### 3.3.1

:bug: Bug Fixes
- Fixed incomplete texts in some buttons.

### 3.3.0

:rocket: New features
- Completed the chatbot according to the latest proposed M46 flow:
  - Updated the possible domains of the question;
  - Reverted the changes in the question flow by re-adding `domainInterest`, `beliefsAndValues`, `socialCloseness` and `positionOfAnswerer` attributes;
  - Added a custom expiration time for nearby questions;
  - Added button to close questions if no answers are received;
  - Removed how helpful was the bot;
  - Added questioner name in the message after sending the answer;
  - Small fixes.

:house: Internal
- Updated project template to version [4.12.5](https://lab.u-hopper.com/devops/ansible/project-template/-/releases/4.12.5).

### 3.2.1

:bug: Bug Fixes
- Cleaned correctly the context after the best answer publish message.

### 3.2.0

:rocket: New features
- Added support for follow-up between requester and responder.
- Added callback to a like message saying to the answerer that the requester likes the provided answer.

:bug: Bug Fixes
- Also delayed messages now are logged correctly.

### 3.1.0

:rocket: New features
- Updated the domain of the question. Reduced the number of attributes from 12 to 8, integrated _sensitive question_ option to question domain.
- Updated the question asking flow/process on telegram according to M46 specification.

:house: Internal
- Updated WeNet Common to version 5.0.1.

### 3.0.0

:rocket: New features

- Added support for task expiration callbacks
- Added support for liking preferred answer
- Added support for showing received answers in batches
- Added support for publishing answers in a preferred way (e.g., Anonymously) or not publishing them

:house: Internal
- Updated project template to version `4.12.2`

## Version 2.*

### 2.2.0

:rocket: New features
- Delayed platform messages so that the in-app message exchange does not get interrupted
- Delayed remind me later messages so that the in-app message exchange does not get interrupted
- Added question in the flow for accepting the answer for publishing questions and best answers in a dedicated telegram channel

:house: Internal
- Updated Chatbot Core to version `1.5.3`

### 2.1.1
- Updated translations
- Updated to Project template version `4.6.1`

### 2.1.0
- Updated translations and added support for translation of badges messages
- Integrated a different conduct message using helper url specific for a pilot if specified in the env vars
- Modified rhe reminder to the user of the attributes he chose when accepting the question

### 2.0.3
- Fixed non-ascii characters are not rendered correctly, json.dumps is applied to user texts and json.loads is applied to platform texts
- Reversed the order of the buttons of the sensitive question
- Reminded to the user the attributes he chose when accepting the question

### 2.0.2
- Updated translations
- Fixed Emojis are not rendered correctly, demojize is applied to user texts and emojize is applied to platform texts
- Added code of conduct message to the start messages
- Updated to project template `4.6.0`

### 2.0.1
- Refactored PendingMessagesJob handling errors and deleting messages that cause these errors

### 2.0.0
- Integrated position data in the question and answer flow
- Integrated sensitive and anonymous information data in the question and answer flow
- Updated to project template `4.5.1`
- Removed submodules and added the libraries as packages:
  - `uhopper-chatbot` version `1.5.2`;
  - `wenet-common` version `3.1.0`;
  - `uhopper-alert` version `2.0.0`;
  - `uhopper-mqtt` version `1.1.0`.
- Updated question flow adding social closeness input in the question creation flow instead of similarity and reason
- Integrated domain data in the question flow
- Integrated domain similarity data in the question flow
- Added environment variables for communityId and maxUsers fields
- Integrated belief and values similarity data in the question flow
- Updated app commands: renamed `/question` into `/ask` and `/answer` into `/questions`
- Integrated rating flow when accepting an answer
- Updated common models to fix service APIs methods for getting all tasks of an application and for an user
- Integrated new structures of the callback messages
- Refactored PendingMessagesJob
- Added `/badges` command to request insights on badge status
- Added a survey promo message in the welcome messages

## Version 1.*

### 1.0.12
- Added sentry integration

### 1.0.11
- Updated common models to hotfix for badges messages

### 1.0.10
- Changed the `project` field in the messages logged into the Wenet platform. Now the `appId` is used, instead of the `PROJECT_NAME` environment variable

### 1.0.9
- Added a new message promoting badges that appears after the login with Wenet
- Updated some translations in spanish and danish

### 1.0.8
- Updated spanish translations

### 1.0.7
- Fixed a bug in showing markdown characters into markdown messages

### 1.0.6
- Updated text of english messages with the new translations provided by Peter

### 1.0.5
- Removed redundant messages.
- Added new logic allowing to clean context once new messages from WeNet are received.
- Added support for "answer picked" notification message.
- Added support for "textual message" notification message.

### 1.0.4
- Several small fixes, such as typos, removal of the command `/profile` and `/report`
- Added the possibility to use an environment variable to set the time to live of the cached locale of each user
- Updated chatbot core to version `1.4.2`
- Added the possibility to use an environment variable to set the name of the log file, as well as the `project` value used in the logging component

### 1.0.3

src/ask_for_help_bot/main.py
from __future__ import absolute_import, annotations

import logging.config
import os
import sentry_sdk

from ask_for_help_bot.handler import AskForHelpHandler
from chatbot_core.translator.translator import Translator
from chatbot_core.v3.connector.social_connectors.telegram_connector import TelegramSocialConnector
from chatbot_core.v3.handler.event_dipatcher import MultiThreadEventDispatcher
from chatbot_core.v3.handler.instance_manager import InstanceManager
from common.logging_config import get_logging_configuration
from uhopper.utils.alert.module import AlertModule
from uhopper.utils.mqtt.handler import MqttSubscriptionHandler
from sentry_sdk.integrations.logging import LoggingIntegration

logging.config.dictConfig(get_logging_configuration(os.getenv("PROJECT_NAME", "wenet-ask-for-help-chatbot")))
logger = logging.getLogger("uhopper.chatbot.wenet-ask-for-help-chatbot")

sentry_logging = LoggingIntegration(
    level=logging.INFO,  # Capture info and above as breadcrumbs
    event_level=logging.ERROR  # Send errors as events
)

sentry_sdk.init(
    # Set traces_sample_rate to 1.0 to capture 100%
    # of transactions for performance monitoring.
    # We recommend adjusting this value in production.
    traces_sample_rate=1.0
)

if __name__ == "__main__":
    topic = os.getenv("MQTT_TOPIC")
    subscriber = MqttSubscriptionHandler(os.getenv("MQTT_HOST"), os.getenv("MQTT_SUBSCRIBER_ID"),
                                         os.getenv("MQTT_USER"), os.getenv("MQTT_PASSWORD"))
    subscriber.add_subscription(topic)
    instance_namespace = os.getenv("INSTANCE_NAMESPACE")
    bot_token = os.getenv("TELEGRAM_KEY")
    connector = TelegramSocialConnector(bot_token)
    alert_module = AlertModule("wenet-ask-for-help-chatbot")
    wenet_instance_url = os.getenv("WENET_INSTANCE_URL")
    wenet_hub_url = os.getenv("WENET_HUB_URL")
    app_id = os.getenv("WENET_APP_ID")
    task_type_id = os.getenv("TASK_TYPE_ID")
    community_id = os.getenv("COMMUNITY_ID")
    max_users = int(os.getenv("MAX_USERS", 5))
    max_answers = int(os.getenv("MAX_ANSWERS", 15))
    expiration_duration = int(os.getenv("EXPIRATION_DURATION", 86400))  # 1 day is 86400 seconds
    nearby_expiration_duration = int(os.getenv("NEARBY_EXPIRATION_DURATION", 7200))  # 2 hours is 7200 seconds
    survey_url = os.getenv("SURVEY_URL")
    helper_url = os.getenv("PILOT_HELPER_URL")
    channel_id = os.getenv("CHANNEL_ID")
    publication_language = os.getenv("PUBLICATION_LANGUAGE", "en")
    wenet_authentication_url = os.getenv("WENET_AUTHENTICATION_URL")
    redirect_url = os.getenv("REDIRECT_URL")
    client_secret = os.getenv("CLIENT_SECRET")
    wenet_authentication_management_url = os.getenv("WENET_AUTHENTICATION_MANAGEMENT_URL")

    translation_folder_path = os.getenv("TRANSLATION_FOLDER_PATH", "../../translations")
    translator = Translator("wenet-ask-for-help", alert_module, translation_folder_path, fallback=False)
    translator.with_language("en", is_default=True, aliases=["en_US", "en_GB"])
    translator.with_language("it", is_default=False, aliases=["it_IT", "it_CH"])
    # translator.with_language("da", is_default=False)
    translator.with_language("mn", is_default=False)
    translator.with_language("es", is_default=False, aliases=["es_ES", "es_PY", "es_AR", "es_MX"])

    handler = AskForHelpHandler(
        instance_namespace=instance_namespace,
        bot_id="wenet-ask-for-help",
        handler_id="wenet-ask-for-help-handler",
        telegram_id=bot_token,
        wenet_instance_url=wenet_instance_url,
        wenet_hub_url=wenet_hub_url,
        app_id=app_id,
        client_secret=client_secret,
        redirect_url=redirect_url,
        wenet_authentication_url=wenet_authentication_url,
        wenet_authentication_management_url=wenet_authentication_management_url,
        task_type_id=task_type_id,
        community_id=community_id,
        max_users=max_users,
        max_answers=max_answers,
        expiration_duration=expiration_duration,
        nearby_expiration_duration=nearby_expiration_duration,
        survey_url=survey_url,
        helper_url=helper_url,
        channel_id=channel_id,
        publication_language=publication_language,
        alert_module=alert_module,
        connector=connector,
        nlp_handler=None,
        translator=translator
    )
    instance_manager = InstanceManager(instance_namespace, subscriber, MultiThreadEventDispatcher())
    instance_manager.with_event_handler(handler)

    try:
        instance_manager.start()
    except KeyboardInterrupt:
        subscriber.disconnect()
